packages/react-router-dom/index.tsx

```diff
@@ -1543,17 +1552,18 @@ export function useFetcher<TData = any>(): FetcherWithComponents<TData> {
     `useFetcher can only be used on routes that contain a unique "id"`
   );
 
-  let [fetcherKey] = React.useState(() => String(++fetcherId));
+  let [_fetcherKey] = React.useState(() => `__${String(++fetcherId)}`);
+  let fetcherKey = key ? key : _fetcherKey;
```

Comment:
Prefer user-specified `key` if available.  Also prefixed our current internal keys with `__` to avoid collisions with user-specified IDs.